### PR TITLE
Catch `matplotlib` crash when running IPython in terminal

### PR DESCRIPTION
**Pull request by strickvl**
## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Describe changes
- Added a check which raises an error if someone tries to run something like `ipython -i examples/whylogs/run.py`
- (Alternative considered: adding these checks to the individual visualisers but that seems like a bad idea.)

### CODE REVIEW

**Review by htahir1 — CHANGES_REQUESTED** [review]
Does this mean IPython would need to be installed or do we have it as a base dependency? If we do then this is fine, if not then we should find another way to detect this with some import handling. I would also advise putting it in the `Environment` class as a static function.

```
def in_terminal() -> bool:
    try:
        from IPython import get_ipython
        if get_ipython().__class__.__name__ == "TerminalInteractiveShell":
            return False
    except ImportError:
        return False
    return True
```

Also, perhaps we need a test for this (Sorry, I seem to be always saying this, and dont want to cause unnecessary overhead but I do feel its important). What is a good way to simulate a IPython environment? Maybe it is not so neccessary but we should at least check manually whether this works.

P.S. Do we need to revise the `[in_notebook](https://github.com/zenml-io/zenml/blob/main/src/zenml/environment.py#L107)` method @jwwwb ? I seem to remember you saying that was flawed too

**Comment by strickvl** [comment]
> with some import handling. I would

Actually we don't have `IPython` as a dependency. So I made the appropriate changes. Thanks for catching that.

Testing is proving a bit more of an issue. I've pushed a test that's currently failing, and I'm trying to `mocker.patch` the relevant object, but to no avail, probably because the same value is being re-imported when it gets called. Will regroup and rework to get it working somehow.

**Comment by htahir1** [comment]
Tbh it isnt worth so much time. lets just merge with your latest commits :-)

File: src/zenml/environment.py
Comment (jwwwb) [inline src/zenml/environment.py:123]: My feedback here would be: Do we need to specifically differentiate between ipython REPL and normal python REPL? I think they should be treated the same to the end user, but the important thing is that they are treated differently from notebook/colab (because of the inlining of plots, etc).

File: src/zenml/environment.py
Comment (jwwwb) [inline src/zenml/environment.py:111]: Specifically I would say that the behavior of this function is incorrect — it returns true when running in an ipython REPL, and I would expect it to be false.

**Review by jwwwb — CHANGES_REQUESTED** [review]

File: src/zenml/environment.py
Comment (strickvl) [reply src/zenml/environment.py:111]: Yep for this that's correct. This was old code prior to my PR which tests for whether it's a notebook in probably not the best way. I'll try to figure out a way to combine these two static methods.

File: src/zenml/visualizers/base_step_visualizer.py
Comment (strickvl) [reply src/zenml/visualizers/base_step_visualizer.py:26]: Ok. I think I was confused because of the proposed solution in the ticket, which suggested targetting things in the `base_step_visualizer`. If we're just going to defer to the behaviour of normal Python, I reckon we'll have to do this in the actual individual visualization steps, if I'm not mistaken?

Comment (jwwwb) [reply src/zenml/visualizers/base_step_visualizer.py:26]: yeah sorry for the confusion @strickvl -- when I wrote that ticket, the code was actually in base_step_visualizer. but since the writing of that ticket, apparently Michael refactored it into the previously non existing environment class. oops. I think the only place it now needs to be changed is in the environment class. we can sync orally in a min too

**Comment by strickvl** [comment]
@jwwwb back over to you. I like this way much more. I actually start having an issue with the whylogs example now (not sure if you do, too?), but that's outside the scope of this ticket I guess.

File: src/zenml/utils/package_utils.py
Comment (htahir1) [inline src/zenml/utils/package_utils.py:18]: I think this might belong in `source_utils` rather than creating another file tbh

**Review by htahir1 — CHANGES_REQUESTED** [review]
I also like this more, but left one thought

File: src/zenml/utils/package_utils.py
Comment (strickvl) [reply src/zenml/utils/package_utils.py:18]: Yes, I agree. Think is, when I add it into `source_utils` I get a circular import since `source_utils` imports `Environment`. Do you think it's better to keep it in `package_utils` or to move the whole function just as a helper in `Environment` that isn't part of our utils?

File: src/zenml/environment.py
Comment (jwwwb) [inline src/zenml/environment.py:114]: aren't we missing an else clause here?

File: src/zenml/utils/package_utils.py
Comment (jwwwb) [reply src/zenml/utils/package_utils.py:18]: my suggestion would be to simply discard this function entirely. When the signature of the function is longer than the body, it probably doesn't warrant a dedicated function.

**Review by jwwwb — CHANGES_REQUESTED** [review]

File: src/zenml/environment.py
Comment (strickvl) [reply src/zenml/environment.py:114]: I don't think so. It'll just return `False` on the final line of the function, as far as I can reason it.

**Review by jwwwb — APPROVED** [review]

**Comment by strickvl** [comment]
@jwwwb I'll wait for tests to pass and then merge it in. Thanks!

**Pull request merged**
